backend/src/apiserver/storage/run_store.go

```diff
@@ -61,32 +61,73 @@ type RunStore struct {
 	time                   util.TimeInterface
 }
 
+// Runs two SQL queries in a transaction to return a list of matching runs, as well as their
+// total_size. The total_size does not reflect the page size, but it does reflect the number of runs
+// matching the supplied filters and resource references.
 func (s *RunStore) ListRuns(
-	filterContext *common.FilterContext, opts *list.Options) ([]*model.Run, string, error) {
-	errorF := func(err error) ([]*model.Run, string, error) {
-		return nil, "", util.NewInternalServerError(err, "Failed to list runs: %v", err)
+	filterContext *common.FilterContext, opts *list.Options) ([]*model.Run, int, string, error) {
+	errorF := func(err error) ([]*model.Run, int, string, error) {
+		return nil, 0, "", util.NewInternalServerError(err, "Failed to list runs: %v", err)
 	}
 
-	sqlBuilder := s.selectRunDetails()
-	// Add filter condition
-	sqlBuilder, err := s.toFilteredQuery(sqlBuilder, filterContext)
+	// SQL for getting the filtered and paginated rows
+	rowsSqlBuilder, err := s.addResourceReferenceToSelect(s.selectRunDetails(), filterContext)
+	if err != nil {
+		return errorF(err)
+	}
+	rowsSql, rowsArgs, err := opts.AddPaginationToSelect(opts.AddFilterToSelect(rowsSqlBuilder)).ToSql()
+	if err != nil {
+		return errorF(err)
+	}
+
+	// SQL for getting total count of the filtered rows
+	// This first query performs a filtered selection over the runs table, it should match
+	// the number of rows returned by the filtered query built above for the actual rows,
+	// but it tries to optimize out the left join, which isn't needed here for the count.
+	countSqlBuilder, err := s.addResourceReferenceToSelect(sq.Select("*").From("run_details"), filterContext)
+	countSqlBuilder = opts.AddFilterToSelect(countSqlBuilder)
+	countSelectBuilder := sq.Select("count(*)").FromSelect(countSqlBuilder, "rows")
```

Comment:
It's not unnecessary, this counts the number of rows returned from the subquery, not in the entire table.

Reply:
Clarified this comment above.  

We want "count(*) from run_details", not "cloud(*) from (select * from run_details)".

Reply:
Sure it would be nice to get to this point, but it won't be simple to do that without refactoring the code in `addResourceReferenceToSelect` and `AddFilterToSelect` to conditionally handle count, making the code brittle.
I'd rather keep the extra layer, which will be optimized by the query execution engine anyway.

Reply:
Performance issues are very hard to debug and fix. Let's please avoid unnecessary complicated SQL statement for the purpose of simplifying code.

Reply:
It's not that simple. The query isn't just `count(*) from jobs where ...`, it's a left join. I might be missing something, but I can't see an easy way to simplify this. The rows of the first subquery need to first join with those of the second, then we count the final result. This is exactly what the query above is doing.

Reply:
What is the join with? It seems the query takes just a filter for the table and the comment says the join is not needed. Even with a join though - the "select count(*) from X join Y" would be faster that with a sub query that may require a full scan.

Why can't we use a direct parameterized SQL query, instead of using the query construction functions?

Reply:
The join is needed in the following two cases:
- Get all runs/jobs: what happens here is there's two left joins, runs -> metrics, then runs -> resource references to augment the runs' rows with their experiments.. etc.
- Get all runs/jobs of a specific experiment (using a resource reference): this is an extra layer of inner joining the filtered runs/jobs from the resource references table on the join result from above.

@IronPan and I discussed this for a while yesterday, and we can see a lot of area for improvements. The code currently is written to optimize for code reuse and cleanliness, at the account of SQL optimization, which I think is good for a first step. I think the next steps should be profiling the code to see if there really are performance issues here, then we can repurpose the code to target speed of execution, perhaps by writing bare string queries, rather than building from reusable objects.

Reply:
PTAL. With @IronPan's latest refactor here, this is now much simpler. I also moved some of the duplicate code in building the list jobs and list runs queries to list_util.